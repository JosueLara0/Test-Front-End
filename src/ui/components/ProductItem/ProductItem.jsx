//* libraries
import { Typography, Button, Box, Rating } from "@mui/material";
//* styles
import { productItemStyles as styles } from "./productItem.styles";

export const ProductItem = ({
  imageUrl,
  installments,
  listPrice,
  price,
  productName,
  stars,
  increment,
}) => {
  return (
    <Box sx={styles.box}>
      {/*//?  Product image  */}
      <Box sx={styles.boxProduct}>
        <Box component="img" src={imageUrl} alt={productName} />

        {listPrice && (
          <Box sx={styles.boxTriangle}>
            <Typography sx={styles.typoOff}>OFF</Typography>
          </Box>
        )}
      </Box>

      {/*//?  Product details  */}
      <Box sx={styles.boxData}>
        <Typography sx={styles.typoName}>{productName}</Typography>

        <Rating name="read-only" value={stars} readOnly sx={styles.rating} />

        {listPrice && (
          <Typography
            sx={styles.typoListPrice}
          >{`de $ ${listPrice}`}</Typography>
        )}

        <Typography sx={styles.typoPrice}>{`por $ ${price}`}</Typography>

        {installments[0] && (
          <Typography
            sx={styles.typoQuantity}
          >{`o en ${installments[0]?.quantity}x de R $ ${installments[0]?.value}`}</Typography>
        )}

        <Button sx={styles.button} onClick={() => increment()}>
          <Typography sx={styles.typoButton}>COMPRAR</Typography>
        </Button>
      </Box>
    </Box>
  );
};
