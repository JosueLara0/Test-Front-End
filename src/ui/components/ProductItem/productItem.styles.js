export const productItemStyles = {
    box: {
        marginRight: '50px',
        display: "flex",
        flexDirection: "column",
        justifyContent: "center",
        alignItems: "center",
    },
    boxProduct: {
        display: "flex"
    },
    boxTriangle: {
        width: "80px",
        height: "70px",
        marginLeft: "-80px",
        background:
            "linear-gradient(to top right, transparent 50%, #F8475F 50%)",
        textAlign: "end",
    },
    typoOff: {
        margin: '12px 8px 0px 0px',
        font: '900 16px Nunito, sans serif',
        color: 'primary.main'
    },
    boxData: {
        textAlign: 'center',
        padding: '7px 20px 14px 20px',
        '&:hover': {
            backgroundColor: "#E6E8EA",
        },
    },
    typoName: {
        font: '600 12px Nunito, sans serif',
        letterSpacing: '0px',
        color: 'secondary.gray'
    },
    rating: {
        "& .MuiRating-iconFilled": {
            color: "error.main",
        },
        "& .MuiRating-iconEmpty": {
            color: "error.main",
        },
    },
    typoListPrice: {
        font: 'normal 12px Nunito, sans serif',
        letterSpacing: '0px',
        color: 'secondary.gray'
    },
    typoPrice: {
        font: 'bold 18px Nunito, sans serif',
        letterSpacing: '0px',
        color: 'secondary.main'
    },
    typoQuantity: {
        font: 'normal 11px Nunito, sans serif',
        letterSpacing: '0px',
        color: 'secondary.gray'
    },
    button: {
        height: '32px',
        width: '125px',
        marginTop: '8px',
        backgroundColor: "secondary.main",
        borderRadius: '5px',
        '&:hover': {
            backgroundColor: "secondary.gray",
        },
    },
    typoButton: {
        font: 'bold 12px Nunito, sans serif',
        letterSpacing: '0px',
    }
};
