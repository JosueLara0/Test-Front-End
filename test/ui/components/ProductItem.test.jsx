//* libraries
import { render, screen } from "@testing-library/react";
//* components
import { ProductItem } from "../../../src/ui/components/ProductItem/ProductItem";

describe("Tests in <ProductItem/>", () => {
  const imageUrl =
    "https://media2.giphy.com/media/oxbNORcXx76F2/giphy.gif?cid=aa030f39gigshg8jcumtz662xsloyek0d7jblzpwqvqnxuhh&rid=giphy.gif&ct=g";
  const installments = [{ quantity: 4, value: 100 }];
  const listPrice = 10;
  const price = 20;
  const productName = "Boots";
  const stars = 4;
  const increment = jest.fn();

  test("should make match with the snapshot", () => {
    const { container } = render(
      <ProductItem
        imageUrl={imageUrl}
        installments={installments}
        listPrice={listPrice}
        price={price}
        productName={productName}
        stars={stars}
        increment={increment}
      />
    );

    expect(container).toMatchSnapshot();
  });

  test("should show img with the right url and alt", () => {
    render(
      <ProductItem
        imageUrl={imageUrl}
        installments={installments}
        listPrice={listPrice}
        price={price}
        productName={productName}
        stars={stars}
        increment={increment}
      />
    );

    const { src, alt } = screen.getByLabelText("productImg");
    expect(src).toBe(imageUrl);
    expect(alt).toBe(productName);
  });

  test("should show OFF text if listPrice exists and his value", () => {
    render(
      <ProductItem
        imageUrl={imageUrl}
        installments={installments}
        listPrice={listPrice}
        price={price}
        productName={productName}
        stars={stars}
        increment={increment}
      />
    );

    const off = screen.getByText("OFF");
    expect(off).toBeTruthy();

    const listPriceValue = screen.getByText(`de $ ${listPrice}`);
    expect(listPriceValue).toBeTruthy();
  });

  test("should show productName, price and rating stars", () => {
    render(
      <ProductItem
        imageUrl={imageUrl}
        installments={installments}
        price={price}
        productName={productName}
        stars={stars}
        increment={increment}
      />
    );

    const productNameValue = screen.getByText(productName);
    expect(productNameValue).toBeTruthy();

    const productPriceValue = screen.getByText(`por $ ${price}`);
    expect(productPriceValue).toBeTruthy();

    const rating = screen.getByLabelText("rating");
    expect(rating).toBeTruthy();
  });

  test("should show quantity and value text if installments exist", () => {
    render(
      <ProductItem
        imageUrl={imageUrl}
        installments={installments}
        listPrice={listPrice}
        price={price}
        productName={productName}
        stars={stars}
        increment={increment}
      />
    );

    const text = screen.getByText(
      `o en ${installments[0]?.quantity}x de R $ ${installments[0]?.value}`
    );
    expect(text).toBeTruthy();
  });
});
