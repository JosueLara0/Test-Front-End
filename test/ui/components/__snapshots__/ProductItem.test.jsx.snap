// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <ProductItem/> should make match with the snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-wkpw2c"
  >
    <div
      class="MuiBox-root css-k008qs"
    >
      <img
        alt="Boots"
        aria-label="productImg"
        class="MuiBox-root css-0"
        src="https://media2.giphy.com/media/oxbNORcXx76F2/giphy.gif?cid=aa030f39gigshg8jcumtz662xsloyek0d7jblzpwqvqnxuhh&rid=giphy.gif&ct=g"
      />
      <div
        class="MuiBox-root css-1dxz2xi"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-11anhiy-MuiTypography-root"
        >
          OFF
        </p>
      </div>
    </div>
    <div
      class="MuiBox-root css-17b89uf"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-19r6slh-MuiTypography-root"
      >
        Boots
      </p>
      <span
        aria-label="rating"
        class="MuiRating-root MuiRating-sizeMedium MuiRating-readyOnly css-1cyf8jp-MuiRating-root"
        role="img"
      >
        <span>
          <span
            class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
        </span>
        <span>
          <span
            class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
        </span>
        <span>
          <span
            class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
        </span>
        <span>
          <span
            class="MuiRating-icon MuiRating-iconFilled css-7qmtgc-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
        </span>
        <span>
          <span
            class="MuiRating-icon MuiRating-iconEmpty css-1c99szj-MuiRating-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
              data-testid="StarBorderIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
              />
            </svg>
          </span>
        </span>
      </span>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1l3kfgd-MuiTypography-root"
      >
        de $ 10
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-9qu0jp-MuiTypography-root"
      >
        por $ 20
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-2wt4ze-MuiTypography-root"
      >
        o en 4x de R $ 100
      </p>
    </div>
  </div>
</div>
`;
